### PR TITLE
[Policies] Document how we use homu

### PR DESCRIPTION
**Pull request by segiddins**

### CODE REVIEW

**Comment by drbrain** [comment]
This doesn't describe how to use homu or link to such a description, just that we're using it.

**Comment by segiddins** [comment]
Nothing else in that document goes into technical specifics, like how to use git tags, and this is in keeping with that

**Comment by hsbt** [comment]
> like how to use git tags

I've never been to look to use homu without bundler and rubygems repository. Please show instructions how to use homu.

**Comment by tenderlove** [comment]
What does Homu buy us?  Why are we using it?  Has the decision to use this already been made, or are we deciding that here?

Also, if that's what we're going to use, shouldn't @drbrain, @hsbt, @evanphx and I be in the list?

**Comment by indirect** [comment]
Okay. In backwards chronological order:

@tenderlove I talked with @drbrain about using Homu in person, and we agreed it would be a good idea, and there have been three follow-up tickets on this repo where we discussed and implemented it. The current maintainers list does not include @drbrain or @evanphx because they have both said they're willing to help if needed, but don't have the time and interest to be active maintainers anymore. It doesn't include @hsbt and you because (as far as I know?) neither one of you has ever been an active maintainer of the standalone RubyGems project.

@hsbt the the workflow is "instead of merging PRs, maintainers comment '@homu r+', and then Homu creates a merge commit and runs the tests, and only advances the master branch if the tests pass".

@segiddins I'd like the docs to explain exactly how we use git branches and tags, too. An oral tradition is a terrible way to transmit required knowledge. 😝

@drbrain so far the policies file has mostly been "what" without "how" or "why". do you think it's reasonable to put all three into the policies doc, or should there be a separate place to explain that in more detail?

**Comment by tenderlove** [comment]
> It doesn't include @hsbt and you because (as far as I know?) neither one of you has ever been an active maintainer of the standalone RubyGems project.

I don't think anyone in the maintainers list meets this criteria (except maybe you now?)

I've never done releases, but I am one of the top 10 committers and currently have commit access.  If contributions can only be done via PR, and homu is the only thing that can merge PRs, then this effectively removes everyone's commit access besides the people in the maintainers list.  Is that the desired result?  ISTM, everyone who has commit access should also be in the maintainers list. I'm not sure why the two lists are different.

**Comment by drbrain** [comment]
When we pulled committer permissions for inactive committers (under the Committer Access policy) we had a definition for what "active" meant.  While this isn't defined in the policies document, we should pull the definition we agreed upon from slack and place it in that section.  So long as a committer meet the definition of "active" they must have the ability to merge PRs.

Also, because RubyGems is integrated in Ruby there is a long-standing agreement that any Ruby committer is automatically a RubyGems committer. (However, active RubyGems maintainers have right of refusal for commits as RubyGems maintainers "own" the direction of RubyGems much like there are owners of date.rb, optparse.rb, etc.)

**Comment by indirect** [comment]
@drbrain thanks, that's very helpful to know! I'll pull the criteria out of the Slack history and add it to the policies document. I think giving @hsbt et. al. commit access to the repo but not Homu reviewer permissions is very much in line with that policy.

@tenderlove the desired result is for no one, including the people on the maintainers list, to be pushing directly to master. What we're trying to do instead is accomplish a minimum of a green-built merge commit and code review from someone other than the PR author. So yes, it effectively removes (direct) commit from everyone, including those on the maintainers list.

Commit bit is a way to empower anyone who actively works on the project, letting them manage branches, issues, and other github-related things, but it is not carte blanche to push to master. In general, I would prefer that only currently active maintainers merge PRs, because they are the ones who have to support and maintain the merged code. (e.g. the "right of refusal for commits" that @drbrain mentions above.)

The maintainers list is people who have agreed to take on time-consuming tasks like ticket triage, PR review, releases, acting as a contact for CoC violations, and other project management tasks. You've contributed a huge amount of useful code to RubyGems in the past, but (historically, anyway) you haven't been active in any of the other aspects of being a maintainer. If you're interested in joining the maintainers team, it would be great to have your help!

**Comment by drbrain** [comment]
I think you missed my intent, if you are a "committer" then you must be able to merge PRs.

RubyGems is a large system with many components, so if a committer's interest is focused to one narrow area it is fine if that is all they wish to review and merge (for example, Windows support).  This allows us to keep subject matter experts interested and active.

Without this ability they become second-class citizens and I can't see why they would want to stick around.  I believe we will be saying "we don't fully value and appreciate your knowledge and expertise, so here is a consolation prize".

**Comment by indirect** [comment]
@drbrain I think our ideas of how this should work are compatible. The problem I'm trying to avoid is "drive-by commits" where a committer who doesn't have to provide support or maintenance for the project pushes new code to master without review and then disappears.

My intent with "active maintainers" vs "committers" is not at all to create second-class citizens. What I want is to have a clear group of people who have agreed to provide time to the project—I  expected that the group of "willing to be asked for a code review anytime" and "has permission to accept PRs" would be the same, but they don't have to be. I'm extremely happy to have people with interest or expertise areas review and merge PRs in that area.

I haven't worked on RubyGems nearly as long as you have, so I'm sure I'm missing context about contributors. Let's actually document who is doing what, so that anyone working on or contributing to RubyGems knows who to ask for review, and when. What changes should I make to the maintainers list to accurately reflect the current state of the project?

**Comment by drbrain** [comment]
tl;dr: I think the maintainers list should be identical to the committers list.  This gives all committers equal standing.  Per the below, we can reclaim commit permissions if committers fall below some activity level.

The goals of these policies should be to maximize inclusiveness.  If a committer follows the policy (uses homu), submits enough PRs and/or makes enough comments to qualify then they should gain all the privileges and rights of anyone else who follows the policy but submits more PRs or makes more review comments.  I don't want to send the message that being able to say "I am a RubyGems maintainer" means "I am willing to take on an unpaid part-time job", especially because it has meant that in the past and it caused burnout of both the original authors and myself.

Earlier we reclaimed commit permissions of inactive committers as a security measure (in case an SSH key gets disclosed to a malicious party, for example).  The remaining committers have shown sufficient commitment to RubyGems so I think it is OK to ask any of them for review, especially if they have touched a relevant area of code (per `git blame`).  If they are not interested or do not have time to review we can reclaim their commit permissions once they've fallen below our activity threshold.

I think we both want to achieve the same thing, that we increase the knowledge shared between contributors and maintainers.  I would prefer to start with the least-possible amount of change to attempt to reach that goal to allow the easiest transition.

Adding homu helps us increase sharing of knowledge as it includes a review step, meaning more people have looked at the changes being made.

I don't feel assigning areas of responsibility for every committer promotes inclusivity (note: I'm quite certain this is an extreme version of "who is doing what" and doesn't reflect your sentiment).  I think we should come to an agreement on what rights and responsibilities "committer" means before visiting this.  In the mean time I think we should start with reasonable judgement (which current committers seem to have) and looking at `git blame` for guidance.

Finally, as we add new policies we should include instructions on how to follow them, but it doesn't have to go here in the policies document.  Without that we aren't helping build each others' knowledge.

**Comment by indirect** [comment]
@drbrain That works for me. Since we already have a policy for inactivity, I think the remaining policy question is what criteria to use when adding new committers. Does it make sense to you to use the MAINTAINERS list to track users who meet the policy criteria? We've historically added new contributors to the GitHub project before they were considered "committers", so they can help with issue triage and the like.

**Comment by drbrain** [comment]
I think if we trust people to triage issues we can trust them to abide by our other policies, so the two sets should be identical.

Since release permissions are controlled through `gem owner` and S3 key any harm a malicious committer causes should be discoverable.

**Comment by segiddins** [comment]
> I think if we trust people to triage issues we can trust them to abide by our other policies, so the two sets should be identical.

I disagree with that. The big issue for me is _responsibility_ -- I think someone (i.e. the person who merges) needs to be responsible for all the code that goes in, not only at the time of merge but also in the aftermath, to fix any bugs the new commits might have caused. And that's a large responsibility that its unreasonable to ask of anyone who's willing to triage issues on an intermittent basis.

**Comment by indirect** [comment]
@drbrain @tenderlove the homu reviewers list (http://homu.io/r/rubygems/rubygems) now mirrors the repository committers list (https://github.com/orgs/rubygems/teams/rubygems). I'll update this PR with a longer explanation of Homu when I get the chance.

**Comment by indirect** [comment]
As I said above, I also think there is a distinction between "maintainers" and "committers", and at some future point I'd like to figure out how to include the difference—but I think using Homu is an improvement entirely by itself. Let's document and start using Homu, and we can have a separate discussion about the possibility of changing merge permissions later.

**Comment by indirect** [comment]
@drbrain @hsbt @tenderlove I've added instructions and linked to more detailed documentation. I'm not sure if I missed anything, because I've been using Homu for a long time—please let me know if there's anything that's still confusing or unclear!

**Comment by drbrain** [comment]
This will get my r+ if it includes the updates to MAINTAINERS.md

**Comment by segiddins** [comment]
@drbrain I think we want maintainers to add themselves to the file, that's what we did originally, that way your preferred name / email are listed.

**Comment by hsbt** [comment]
@indirect I understood usage of home and confirmed to http://homu.io/r/rubygems/rubygems .

I only focused to guarantee compatibility and integrity of rubygems and ruby core. I'm not standalone maintainer probably. but I hope that we get sustainable development of rubygems and ruby. I proud to maintain rubygems.

File: POLICIES.rdoc [hunk 1]
```diff
@@ -1,3 +1,16 @@
+== Pull Requests
+
+Contributions to RubyGems are made via GitHub pull requests, which must be
+approved by a project committer other than the author. To approve a PR, a
+maintainer can leave a comment including the text "@homu r+", indicating that
+they have reviewed the PR and approve it. [Homu](http://homu.io) will then
+automatically create a merge commit, test the merge, and land the PR if the
+merge commit passes the tests.
+
+This process guarantees that our release branches are always be green, and
```
Comment (duckinator) [inline POLICIES.rdoc]: It'd be clearer to explicitly say what a branch being "green" means.

E.g., "This process guarantees that our release branches never have failing tests, [...]"

**Comment by duckinator** [comment]
Aside from my suggestion (https://github.com/rubygems/rubygems/pull/1518#discussion_r55774555), is there anything blocking this from being merged?

**Comment by indirect** [comment]
Merged manually and fixed @duckinator's note, since I don't have push on this PR.

**Comment by indirect** [comment]
Oops. Turns out I did have push on this PR, but I didn't realize it because it was from a fork rather than this repo. :grimacing:
